src/Helpers/SAMLUserGroupMapper.php

```diff
@@ -0,0 +1,135 @@
+
+    /**
+     * Allow addition of member to a manually-created group which does not exist on IdP
+     */
+    private static bool $allow_manual_group = false;
+
+    /**
+     * Check if group claims field is set and assigns member to group
+     *
+     * @param [] $attributes
+     * @param Member $member
+     * @return Member
+     */
+    public function map($attributes, $member): Member
+    {
+        $logger = Injector::inst()->get(LoggerInterface::class);
+        $groupClaimsField = $this->config()->get('group_claims_field');
+        $groupMap = $this->config()->get('group_map');
+
+        $logger->info(sprintf('IdP Attributes: %s', json_encode($attributes, JSON_THROW_ON_ERROR)));
+
+        // Check if group mapping config exists
+        if (count($groupMap) <= 0) {
+            return $member;
+        }
+
+        // Remove member from any group (except if manual group is allowed) before syncing
+        if ($this->config()->get('allow_manual_group')) {
+            $this->removeMemberFromGroups($member, true);
+        } else {
+            $this->removeMemberFromGroups($member);
+        }
+
+        // Get groups from SAML response
+        if (!array_key_exists($groupClaimsField, $attributes)) {
+            return $member;
+        }
+
+        $groups = $attributes[$groupClaimsField];
+
+        // Check that group claims field has sent through from provider
+        if (!isset($groups)) {
+            return $member;
+        }
+
+        foreach ($groups as $groupID) {
+            // Check that group is a valid group with group map
+            if (!array_key_exists($groupID, $groupMap)) {
+                continue;
+            }
+            $groupTitle = $groupMap[$groupID];
+
+            // Get Group object by Title
+            $group = DataObject::get_one(Group::class, [
+                '"Group"."Title"' => $groupTitle
+            ]);
+
+            // Create group if it doesn't exist yet
+            if (!$group) {
+                $group = new Group();
+                $group->Title = $groupTitle;
+                $group->write();
+            }
+
+            // Add member to the group
+            $group->Members()->add($member);
+        }
+
+        return $member;
+    }
+
+    /**
+     * Remove the member from current CMS groups except for manual override
+     */
+    protected function removeMemberFromGroups(Member &$member, bool $allowManualGroup = false)
+    {
+        if (!$member) {
+            return false;
+        }
+
+        // Remove all groups associated with this member
+        if (!$allowManualGroup) {
+            $member->Groups()->removeAll();
+            return false;
+        }
+
+        $groupMap = $this->config()->get('group_map');
+
+        // Check if group mapping exists
+        if (count($groupMap) <= 0) {
+            return false;
+        }
+
+        // loop through defined group map and remove member
+        foreach ($groupMap as $id => $groupTitle) {
```

Comment:
Is it not easier to loop over `$member->Groups()` instead of looking up groups one by one?

Reply:
It's easier still to use `->RemoveAll()` :)